client/me/purchases/utils.ts

```diff
@@ -7,12 +7,16 @@ import {
 	isPaidWithCreditCard,
 } from 'calypso/lib/purchases';
 import { addPaymentMethod, changePaymentMethod, addNewPaymentMethod } from './paths';
+import type { Purchase } from 'calypso/lib/purchases/types';
 
-function isDataLoading( props ) {
+export function isDataLoading( props: {
+	hasLoadedSites: boolean;
+	hasLoadedUserPurchasesFromServer: boolean;
+} ): boolean {
 	return ! props.hasLoadedSites || ! props.hasLoadedUserPurchasesFromServer;
 }
 
-function canEditPaymentDetails( purchase ) {
+export function canEditPaymentDetails( purchase: Purchase ) {
```

Comment:
Seems like this would expect a boolean as a response. Should it be included as the return type here?

Reply:
A good question! It does have an _implicit_ boolean return value, but TypeScript can infer this and got lazy 😅 . Unlike PHP, inferred types are just as valid and strict as explicit types, so if if you tried to, say, use string manipulation on the result of this function, you wouldn't be able to compile it (eg: "cannot use this boolean as a string"). 

The advantage of adding an explicit return type would be that if this function currently or in the future tried to return something different, a type error would appear here (eg: "this function returns a boolean but it can return a string"). By leaving the return type off, we're telling TypeScript, "the return type of this function is whatever it can return". If we modified it to sometimes return a string, the implicit return type would become `boolean|string` and TypeScript would make sure that everywhere that called it was ok with having a string sometimes.

It's a trade-off either way. [This SO answer](https://stackoverflow.com/questions/70001511/why-specify-function-return-types) does a pretty good job of explaining it. I generally prefer adding return types but when converting existing code sometimes I don't as an expedient. That said, I may go back and add them here.

Reply:
Here's an example of the inferred types at work. When converting this function to TS maybe I don't know what it's supposed to return, so I can just let TS figure it out:

<img width="691" alt="Screenshot 2024-12-12 at 3 39 37 PM" src="https://github.com/user-attachments/assets/ebd71422-775a-4611-abf8-b884a9df02d2" />